File: api/routes/analytics.py
"""Analytics API routes — training statistics, trends, and muscle distribution."""
import uuid
import logging
from datetime import date
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies.database import get_db
from api.schemas import (
    AnalyticsOverviewResponse,
    WeeklyAnalyticsPoint,
    ExerciseTrendResponse,
    MuscleDistributionItem,
)
from services.analytics_service import AnalyticsService, AnalyticsError

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/users/{user_id}/analytics", tags=["analytics"])


@router.get("/overview", response_model=AnalyticsOverviewResponse)
async def get_overview(
    user_id: uuid.UUID,
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """Get training overview statistics. Defaults to last 28 days."""
    svc = AnalyticsService(db)
    try:
        return await svc.get_overview(user_id, date_from=date_from, date_to=date_to)
    except AnalyticsError as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/weekly", response_model=List[WeeklyAnalyticsPoint])
async def get_weekly(
    user_id: uuid.UUID,
    weeks: int = Query(8, ge=1, le=52),
    db: AsyncSession = Depends(get_db),
):
    """Get weekly aggregated training data."""
    svc = AnalyticsService(db)
    try:
        return await svc.get_weekly_activity(user_id, weeks=weeks)
    except AnalyticsError as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/exercises/{exercise_id}", response_model=ExerciseTrendResponse)
async def get_exercise_trend(
    user_id: uuid.UUID,
    exercise_id: uuid.UUID,
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """Get performance trend for a specific exercise."""
    svc = AnalyticsService(db)
    try:
        return await svc.get_exercise_trend(user_id, exercise_id, date_from=date_from, date_to=date_to)
    except AnalyticsError as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)


@router.get("/muscles", response_model=List[MuscleDistributionItem])
async def get_muscle_distribution(
    user_id: uuid.UUID,
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    """Get training distribution by muscle group."""
    svc = AnalyticsService(db)
    try:
        return await svc.get_muscle_distribution(user_id, date_from=date_from, date_to=date_to)
    except AnalyticsError as e:
        raise HTTPException(status_code=e.status_code, detail=e.message)

File: frontend/lib/api/analytics.ts
import { apiFetch } from "./client"

export interface AnalyticsOverview {
  period: { start: string; end: string }
  completed_workouts: number
  total_duration_seconds: number
  average_duration_seconds: number
  completed_sets: number
  total_reps: number
  total_volume: number
  average_rpe: number | null
  current_streak_days: number
  longest_streak_days: number
  data_quality: { has_weight_data: boolean; has_rpe_data: boolean }
}

export interface WeeklyAnalyticsPoint {
  week_start: string
  completed_workouts: number
  completed_sets: number
  total_reps: number
  total_volume: number
  average_rpe: number | null
  total_duration_seconds: number
}

export interface ExerciseTrendResponse {
  exercise: { id: string; name: string }
  summary: {
    session_count: number; completed_set_count: number; total_reps: number
    total_volume: number; max_weight: number | null; best_set_reps: number | null
    average_rpe: number | null; last_performed_at: string | null
  }
  trend: { date: string; max_weight: number | null; total_volume: number; total_reps: number; average_rpe: number | null }[]
  trend_direction: string
}

export interface MuscleDistributionItem {
  primary_muscle: string
  completed_sets: number
  total_reps: number
  total_volume: number
  percentage_by_sets: number
}

export async function getOverview(userId: string, from?: string, to?: string): Promise<AnalyticsOverview> {
  const params = new URLSearchParams()
  if (from) params.set("date_from", from)
  if (to) params.set("date_to", to)
  const qs = params.toString() ? "?" + params.toString() : ""
  return apiFetch<AnalyticsOverview>(`/users/${userId}/analytics/overview${qs}`)
}

export async function getWeeklyActivity(userId: string, weeks = 8): Promise<WeeklyAnalyticsPoint[]> {
  return apiFetch<WeeklyAnalyticsPoint[]>(`/users/${userId}/analytics/weekly?weeks=${weeks}`)
}

export async function getExerciseTrend(userId: string, exerciseId: string): Promise<ExerciseTrendResponse> {
  return apiFetch<ExerciseTrendResponse>(`/users/${userId}/analytics/exercises/${exerciseId}`)
}

export async function getMuscleDistribution(userId: string): Promise<MuscleDistributionItem[]> {
  return apiFetch<MuscleDistributionItem[]>(`/users/${userId}/analytics/muscles`)
}
